Graph API: Operation Graph creation and interface

amberhassaan opened the pull request:
graph construction and interfaces.

amberhassaan committed:
WIP: graph creation and interface

amberhassaan committed:
WIP: add a test for op graph

amberhassaan committed:
Merge remote-tracking branch 'origin/develop' into amber/graph

amberhassaan committed:
WIP: tests for op graph

amberhassaan committed:
initial test works

amberhassaan committed:
Cleanup

amberhassaan committed:
formatting fixes

junliume commented:
@amberhassaan we are having some build issues with this PR, please fix it. Thanks!

scerzh commented on src/include/miopen/graphapi/opgraph.h:
```diff
@@ -0,0 +1,274 @@
+
+#include <algorithm>
+#include <vector>
+#include <unordered_map>
```
```suggestion
#include <algorithm>
#include <unordered_map>
#include <vector>
```
Sort this list, please

amberhassaan replied:
done

scerzh commented on src/include/miopen/graphapi/opgraph.h:
```diff
@@ -0,0 +1,274 @@
+struct OpNode
+{
+
+    using Edge = std::pair<OpNode*, Tensor*>;
```
```suggestion
struct OpNode
{
    virtual ~OpNode() = default;

    using Edge = std::pair<OpNode*, Tensor*>;
```
A polymorphic base should have a virtual destructor.

amberhassaan replied:
done, in cpp file.

scerzh commented on src/include/miopen/graphapi/opgraph.h:
```diff
@@ -0,0 +1,274 @@
+
+    void addNodes(std::vector<OpNode*>&& nodes) { mNodes = std::move(nodes); }
+
+    void addEdge(OpNode* src, Tensor* tens_ptr, OpNode* dst)
+    {
+        assert(src);
+        assert(dst);
+        src->addOutEdge(dst, tens_ptr);
+        dst->addInEdge(src, tens_ptr);
+    }
```
Should this member function be static as it just links the nodes without accessing `*this`?

amberhassaan replied:
no, we might end up needing it to not be static, e.g., update some graph state. We might also start keeping the neighbor info in the `Graph` itself.

scerzh reviewed (commented):
Some small comments

scerzh commented on src/include/miopen/graphapi/opgraph.h:
```diff
@@ -0,0 +1,274 @@
+#pragma once
```
```suggestion
/*******************************************************************************
 *
 * MIT License
 *
 * Copyright (c) 2023 Advanced Micro Devices, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 *******************************************************************************/
 
 #pragma once
```
License is missing.

amberhassaan replied:
done

scerzh commented on src/include/miopen/graphapi/opgraph.h:
```diff
@@ -0,0 +1,274 @@
+#pragma once
+
+#include <miopen/graphapi/graphapi_tensor.hpp>
```
I'm going to rename this file
miopen/graphapi/graphapi_tensor.hpp => miopen/graphapi/tensor.hpp

amberhassaan replied:
That's fine. It's only one line to update.

amberhassaan committed:
Merge remote-tracking branch 'origin/develop' into amber/graph

amberhassaan committed:
address comments

amberhassaan committed:
Merge remote-tracking branch 'origin/develop' into amber/graph

amberhassaan committed:
fix build

msaudulhassan commented on src/include/miopen/graphapi/opgraph.h:
```diff
@@ -0,0 +1,223 @@
+public:
+    bool hasNode(OpNode* node) const { return internal::contains(mNodes, node); }
+
+    void addNode(OpNode* node)
+    {
+        assert(!hasNode(node));
+        mNodes.emplace_back(node);
+    }
```
Would the assert impact recursive operations?

amberhassaan replied:
not sure what you mean. Do you mean a self cycle?

msaudulhassan replied:
Yes, but I believe a self-cycle can be represented by smtg like 
``` DummyNode top{"top", {&t_in}, {&t_in}}; ```

But what about an unrolled self-cycle, where the same node would occur again and again?

amberhassaan replied:
We will have a `Validate` step for graph in a future PR that will detect self-cycles. AFA having the same operation multiple times in the graph, that is OK as long as they are different objects with different tensors as inputs and outputs, e.g., 
```
DummyNode one{"scale", x, y};
DummyNode two{"scale", y, z};
DummyNode three{"scale", z, out};
```

msaudulhassan replied:
Like
![image](https://github.com/ROCm/MIOpen/assets/68208941/1af85f08-9591-4f66-9cb0-8b1ee883532b)

amberhassaan replied:
left side is not OK. Right side is OK, so the user will have to manually unroll such a computation.

scerzh reviewed (approved):
LGTM

CAHEK7 commented on src/graphapi/opgraph.cpp:
```diff
@@ -0,0 +1,107 @@
+        for(Tensor* o : n->getOutTensors())
+        {
+            auto [iter, _ignore] =
+                e_map.emplace(o, EdgeInfo{}); // add empty EdgeInfo if not present
```
I'd use `try_emplace`, since emplace has always constructed `std::pair<key, val>` even if an object exists and ne emplace will be done.

CAHEK7 commented on src/graphapi/opgraph.cpp:
```diff
@@ -0,0 +1,107 @@
+        for(Tensor* i : n->getInTensors())
+        {
+            auto [iter, _ignore] =
+                e_map.emplace(i, EdgeInfo{}); // add empty EdgeInfo if not present
```
I'd use `try_emplace`, since emplace has always constructed `std::pair<key, val>` even if an object exists and ne emplace will be done.

CAHEK7 commented on src/graphapi/opgraph.cpp:
```diff
@@ -0,0 +1,107 @@
+        }
+        else if(edge_info.mDests.empty())
+        {
+            // tens_ptr is a non-virtual output tensor
+            assert(!tens_ptr->isVirtual());
```
This one is critical. The other comments are less critical and I don't care much about them.

It can (and will) be virtual.
Forward fp8-mha uses virtual output nodes to reflect that it works in inference mode.

Moreover we must be able to match one algorithm to the both cases. Probably it can be worked around by advertising two different signatures.

amberhassaan replied:
Yeah, I did realize after you mentioned it that an output tensor can be virtual. I will remove the `assert` to allow it. 
In general, `virtual/non-virtual` will not be part of the graph signature (as of now at least).

CAHEK7 commented on src/include/miopen/graphapi/opgraph.h:
```diff
@@ -0,0 +1,223 @@
+#include <miopen/graphapi/tensor.hpp>
+
+#include <algorithm>
+#include <unordered_map>
```
I'd suggest moving this include into cpp file, since we actually don't want to include `<unordered_map>` everywhere we use graph.

amberhassaan replied:
good catch. I started out with everything in the header, being lazy.

CAHEK7 commented on src/include/miopen/graphapi/opgraph.h:
```diff
@@ -0,0 +1,223 @@
+
+    virtual std::vector<Tensor*> getOutTensors() const = 0;
+
+    const auto& iterateInEdges() const { return mInEdges; }
+
+    const auto& iterateOutEdges() const { return mOutEdges; }
```
It sounds like a `visitor`, but actually it's a `getter`.

amberhassaan replied:
Can rename to `get`.

CAHEK7 commented on src/include/miopen/graphapi/opgraph.h:
```diff
@@ -0,0 +1,223 @@
+
+    void addOutEdge(OpNode* dst, Tensor* tens_ptr)
+    {
+        assert(dst);
```
Why only `dst` is checked here? Can `tens_ptr` be null?
The same queston is for the `addInEdge()` method and `OpGraph::hadEdge/addEdge`

amberhassaan replied:
Will add the second assert. I missed it because initially I wasn't storing `tens_ptr` with the `neighbor node`.

CAHEK7 commented on src/include/miopen/graphapi/opgraph.h:
```diff
@@ -0,0 +1,223 @@
+private:
+    friend class OpGraphBuilder;
+
+    void addNodes(std::vector<OpNode*>&& nodes) { mNodes = std::move(nodes); }
```
It looks like `setNodes`, since it replaces the previous state.

amberhassaan replied:
yeah, I need a better name. Perhaps `moveNodes` because `setNodes` doesn't quite cut it. I hate setters in general.

CAHEK7 commented on test/gtest/graphapi_opgraph.cpp:
```diff
@@ -0,0 +1,122 @@
+{
+    assert(name.size() <= sizeof(int64_t));
+    int64_t id = 0;
+    std::copy(name.begin(), name.end(), reinterpret_cast<char*>(&id));
```
Formally speaking it must be a plain `std::memcpy` for such kind of tricks.

amberhassaan replied:
`std::copy`  in C++ == `memcpy` in C for contiguous objects.

CAHEK7 replied:
Not exactly that, if we are talking about type punning.
`std::copy` is a typed operation and an stl algorithm - it's up to the library implementation to decide how to perform data copy.
Luckly we reinterpret it as a `char*`, that is probably safe.

Anyway, that is much safer:
```suggestion
    std::copy_n(name.begin(), sizeof(id), reinterpret_cast<char*>(&id));
```
In the worst case we truncate input name, but we will never cross dst buffer boundary and corrupt any data.

amberhassaan replied:
yes, `copy_n` is a better idea.

CAHEK7 commented on test/gtest/graphapi_opgraph.cpp:
```diff
@@ -0,0 +1,122 @@
+template <bool IsVirtual>
+gr::Tensor makeDummyTensor(std::string_view name)
+{
+    assert(name.size() <= sizeof(int64_t));
```
I don't think it should be an assertion. I'd prefer explicit check.

amberhassaan replied:
I'm not entire sure about this trick I put in. Perhaps we should just name the tensors with a single `char` and upcast that to `int64_t`. I thought we have 8 bytes, why not allow multiple characters.

CAHEK7 commented on src/include/miopen/graphapi/opgraph.h:
```diff
@@ -0,0 +1,223 @@
+        std::vector<OpNode*> mDests{};
+    };
+
+    OpGraph build();
```
You can't call this method multiple times and it's not obvious.
Either it should be obvious and bullet proof, or `mNodes` must not be moved.

amberhassaan replied:
OK, so I forgot to make it an `r-value` method. That's the general pattern we will have with `builders` that `build()` will be an r-value method that consumes the calling builder object.

CAHEK7 commented on test/gtest/graphapi_opgraph.cpp:
```diff
@@ -0,0 +1,122 @@
+    graph_builder.addNode(&right);
+    graph_builder.addNode(&bottom);
+
+    gr::OpGraph graph = std::move(graph_builder).build();
```
This `std::move` looks a bit odd. Could you explain why you use it?
I can make multiple assumptions and all of them do not look good, but probably I'm missing something.

amberhassaan replied:
`build()` will be an r-value method.

CAHEK7 reviewed (changes requested):
There is one critical comment.
The others are up to you to fix it or to ignore.

amberhassaan committed:
address comments

amberhassaan committed:
Merge remote-tracking branch 'origin/develop' into amber/graph

amberhassaan committed:
combine duplication of OpNode and fix up Convolution Operation classes

amberhassaan committed:
fix formatting

amberhassaan commented:
@scerzh : please re-review changes to Convolution Operation classes that inherit from `OpNode`. I removed the duplicate `OpNode` and added `signName()` virtual method. 
@CAHEK7 : I should have addressed all the comments unless I missed something.

amberhassaan committed:
use `copy_n` instead of `copy`.

Co-authored-by: Alex Eremin <CAHEK7@yandex.ru>

scerzh reviewed (approved):
LGTM

atamazov commented on test/gtest/graphapi_opgraph.cpp:
```diff
@@ -0,0 +1,122 @@
+{
+    assert(name.size() <= sizeof(int64_t));
+    int64_t id = 0;
+    std::copy_n(name.begin(), sizeof(id), reinterpret_cast<char*>(&id));
```
This may try to read `name` past the end of it.
```suggestion
    std::copy_n(name.begin(), name.size(), reinterpret_cast<char*>(&id));
```

CAHEK7 replied:
> This may try to read `name` past the end of it.

May try to write past the `sizeof(id)`.
It must be something like:
```C++
std::copy_n(name.begin(), std::min(name.size(), sizeof(id)), reinterpret_cast<char*>(&id));
```

amberhassaan replied:
yes, that's what I ended up doing in the next PR. Will bring the fix back here today.

atamazov replied:
[resolved]

atamazov replied:
@CAHEK7 
> May try to write past the `sizeof(id)`.

No, because of `assert(name.size() <= sizeof(int64_t))`. 

[Off-topic] The code that relies on assert actually implies:
- The programmer guarantees that the condition will always be met.
- If condition is not met, the program may work incorrectly

If we afraid that this may happen in release builds (i.e. the programmer can not actually guarantee that the condition will always be met), then assert should be changed to MIOPEN_THROW (but please consider performance penalty).

It must be something like:
> 
> ```c++
> std::copy_n(name.begin(), std::min(name.size(), sizeof(id)), reinterpret_cast<char*>(&id));
> ```

It's sad to say that, but this solution may camouflage the actual errors. This code will work properly only if name fits into id. Otherwise the loss of information will happen and that may lead to incorrect results. Thanks!

CAHEK7 replied:
> @CAHEK7
> 
> > May try to write past the `sizeof(id)`.
> 
> No, because of `assert(name.size() <= sizeof(int64_t))`.

@atamazov 
No. There is no assertion in the release. And that's 100% buffer overrun and assertion is not enough to properly handle this case. It's not just "incorrectly".

And there is no problem with truncation, but it's better to avoid this. Moreover, it's a test itself, it's not a library. THe only thing I may suggest is to put it into the logs, but it's a test, not the library and nobody sees it.

atamazov replied:
@CAHEK7 truncation of the name may spoil uniqueness of the name. When you write `std::min(name.size(), sizeof(id))` you implicitly say that truncation is Ok.

amberhassaan replied:
yes, we are saying that names greater than 8 characters are not allowed.

atamazov replied:
@amberhassaan perhaps let's just fail the test when `(name.size() > sizeof(int64_t))`?

amberhassaan replied:
agree. Fix coming in #2855: in 6cc9d8c

atamazov commented on test/gtest/graphapi_opgraph.cpp:
```diff
@@ -0,0 +1,122 @@
+    DummyNode right{"right", {&t_b}, {&t_d}};
+    DummyNode bottom{"bottom", {&t_c, &t_d}, {&t_out}};
+
+    graph_builder.addNode(&top);
+    graph_builder.addNode(&left);
+    graph_builder.addNode(&right);
+    graph_builder.addNode(&bottom);
```
[Recommendation] Maybe it is worth adding the same nodes but in different order and then check the same assertions? Up to you.

amberhassaan replied:
I don't think the order matters but I ended up adding more tests to this in #2855 and caught a bug too.

atamazov replied:
[resolved]

atamazov commented on src/graphapi/opgraph.cpp:
```diff
@@ -0,0 +1,109 @@
+        {
+
+            // tens_ptr is a non-virtual input tensor
+            assert(!tens_ptr->isVirtual()); // tensor pointer can't be virtual
```
```C++
... // tensor pointer can't be virtual
```
Yep, this looks like a misconfigured graph. But do we \<plan to\> guarantee that this condition is met (i.e. have the same check somewhere else)? If not, then I would highly recommend converting this to MIOPEN_THROW.

CAHEK7 replied:
Seems it must not be checked.
Input tensors can be some sort of virtual.
For example batchnorm operation consumes a lot of input tensors, but some of them are optional.

amberhassaan replied:
I will have to think about it. What does it mean for inputs to be virtual? I thought inputs must be provided or you can't compute a valid result but you're right there might be cases when inputs are optional and a valid output can still be produced.

amberhassaan replied:
I've taken the check out for now. Both input and output tensors can be virtual.

CAHEK7 replied:
> I will have to think about it. What does it mean for inputs to be virtual? I thought inputs must be provided or you can't compute a valid result but you're right there might be cases when inputs are optional and a valid output can still be produced.

The primitive just ignores some inputs, that's it. Either the input can be some precalculated value and primitive calculates it by itself, or it just work in another mode without those tensors.
There are two options:
1) the user does not provide those tensors to the primitive at all - we won't see any input virtual tensors, but we have to feed nullpointers internally
2) the user explicitly marks those tensors virtual

amberhassaan replied:
yeah, it's possible that some operations don't need all inputs, in which case the user must mark them `virtual` in my opinion.

atamazov replied:
[resolved]

atamazov commented on src/graphapi/opgraph.cpp:
```diff
@@ -0,0 +1,109 @@
+        else
+        {
+            // both can't be true at the same time
+            assert(!(edge_info.mSrc == nullptr && edge_info.mDests.empty()));
```
[Quality] This condition is also checked at lines 83 and 95, which means code duplication (albeit the very local one). Please consider refactoring this.

CAHEK7 replied:
This is actually "always false" `else` case.
We can get to the else case if `!(!src & !dst) & !src & !dst` which is `(src | dst) & !src & !dst` which is `(src & !src & !dst) | (dst & !src & !dst)` and as `x & !x == false` we have `false | false` `else` case.

The assert is not required at all and it's just a dead code. `Else` case is impossible.

amberhassaan replied:
`else` is possible but I moved the assert in else up to a `MIOPEN_THROW_IF`.

CAHEK7 replied:
> `else` is possible but I moved the assert in else up to a `MIOPEN_THROW_IF`.

Else is not possible for given two variables. If you don't believe in math, that's an experiment:
https://godbolt.org/z/9EKnb6b15

But top-level `MIOPEN_THROW_IF` is a good solution.

atamazov replied:
[resolved]

amberhassaan replied:
ok, I see it now. The second `else if` masks the error condition in the last `else`, but now we can check for it up with the `MIOPEN_THROW_IF`.

atamazov commented on src/graphapi/opgraph.cpp:
```diff
@@ -0,0 +1,109 @@
+        else
+        {
+            // both can't be true at the same time
+            assert(!(edge_info.mSrc == nullptr && edge_info.mDests.empty()));
```
[R] Maybe it is also worth changing from assert to MIOPEN_THROW, please consider this, thanks.

CAHEK7 replied:
Except the fact that `else` case is impossible.

atamazov replied:
[resolved]

atamazov commented on src/graphapi/opgraph.cpp:
```diff
@@ -0,0 +1,109 @@
+
+OpNode::~OpNode() = default;
+
+OpGraph OpGraphBuilder::build() &&
```
Maybe I am missing something, but -- can you please explain a little why we want to allow this method for rvalues only?

CAHEK7 replied:
It's required to make `build` a bit more explicit about its state invalidation nature.
You can't call `build` twice.

Though it doesn't force anything, because there is no determined `moved from` state by standard.

amberhassaan replied:
@atamazov : the idea is that we use `builder` as a temporary object to construct another object and then throw it away. I am not a fan but sometimes it's necessary when you cannot create the target object in one shot. Imagine for examples if graph nodes were being read from the file or network, then builder can capture them and build the graph when `build()` is called. 

In this case, it's partly due to how the Backend C API is, where each attribute is initialized one at a time.

amberhassaan replied:
Let's have a discussion but not block this PR. We are going to use the `builder pattern` for this API as it's a logical fit. May be we can have the `builder` classes inherit from a common `base class` that holds a `flag` that is set when the object is `moved` from or `consumed`.

atamazov reviewed (changes requested):
Good size of PR 👍

amberhassaan committed:
Merge remote-tracking branch 'origin/develop' into amber/graph

amberhassaan committed:
address comments

amberhassaan commented:
@atamazov , @CAHEK7 : I'd like to merge this PR asap. Let's address the remaining comments and follow up work.

amberhassaan committed:
fix copy_n

atamazov commented:
@amberhassaan 🟢 I am not against merging this PR as is, provided that all of the review comments will be resolved in the near future. 
- If the comment is resolved during the discussion (i.e. resolving does not involve changes in the code), this is Ok.
- If you are going to provide fixes in any future PR, please update the relevant review thread(s) with links to the relevant code snippets or commits (to the PR(s) that contain the actual fixes). That would save the time of the reviewers.

amberhassaan commented:
@junliume : please merge this PR if it passes CI.
